File: app/src/main/java/app/we/go/emojidraw/ThisApplication.java
package app.we.go.emojidraw;

import android.app.Application;
import android.content.Context;

import app.we.go.emojidraw.arch.di.ApplicationComponent;
import app.we.go.emojidraw.arch.di.ApplicationModule;
import app.we.go.emojidraw.arch.di.DaggerApplicationComponent;


public class ThisApplication extends Application {

    private ApplicationComponent component;

    @Override
    public void onCreate() {
        super.onCreate();
        component = createComponent();
    }

    protected ApplicationComponent createComponent() {
        return DaggerApplicationComponent.builder()
                .applicationModule(new ApplicationModule(this))
                .build();
    }


    public static ApplicationComponent getComponent(Context context) {
     return ((ThisApplication) context.getApplicationContext()).component;
    }



}

File: app/src/main/java/app/we/go/emojidraw/ThisApplication.kt
package app.we.go.emojidraw

import android.app.Application
import android.content.Context
import app.we.go.emojidraw.arch.di.ApplicationComponent
import app.we.go.emojidraw.arch.di.ApplicationModule
import app.we.go.emojidraw.arch.di.DaggerApplicationComponent
import timber.log.Timber
import timber.log.Timber.DebugTree


open class ThisApplication: Application() {

    private lateinit var component: ApplicationComponent

    companion object {
        fun getComponent(context: Context): ApplicationComponent {
            return (context.applicationContext as ThisApplication).component
        }
    }

    override fun onCreate() {
        super.onCreate()
        if (BuildConfig.DEBUG) {
            Timber.plant(DebugTree())
        }
        component = createComponent()

    }

    protected open fun createComponent(): ApplicationComponent {
        return DaggerApplicationComponent.builder()
                .applicationModule(ApplicationModule(this))
                .build()
    }



}